### PR TITLE
lightning: auto resume from recoverable error

### LINKED ISSUE #33363: lightning: auto resume from connection error
## Enhancement

Don't record connection error to checkpoint, so that Lightning can resume automatically. Connection error is usually recoverable error. It doesn't cause any data corruption. So it is safe to resume Lightning without any other operations.

### PR DESCRIPTION
**Pull request by buchuitoudegou**
### What problem does this PR solve?

Issue Number: close #33363 

Problem Summary:

### What is changed and how it works?
Ignoring recoverable errors before writing checkpoints.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/restore/restore.go [hunk 1]
```diff
@@ -975,6 +975,11 @@ func (rc *Controller) saveStatusCheckpoint(ctx context.Context, tableName string
 	case !common.IsContextCanceledError(err):
 		merger.SetInvalid()
 		rc.errorSummaries.record(tableName, err, statusIfSucceed)
+	case utils.IsRetryableError(err):
```
Comment (gozssky) [inline br/pkg/lightning/restore/restore.go:978]: It seems retryable error doesn't contain "connection refused"?

Comment (buchuitoudegou) [reply br/pkg/lightning/restore/restore.go]: Try fixing it with `MessageIsRetryableStorageError`.

**Comment by glorv** [comment]
Please add a unit test

File: br/pkg/lightning/restore/restore.go [hunk 2]
```diff
@@ -975,6 +975,11 @@ func (rc *Controller) saveStatusCheckpoint(ctx context.Context, tableName string
 	case !common.IsContextCanceledError(err):
 		merger.SetInvalid()
 		rc.errorSummaries.record(tableName, err, statusIfSucceed)
+	case utils.IsRetryableError(err), utils.MessageIsRetryableStorageError(err.Error()):
```
Comment (glorv) [inline br/pkg/lightning/restore/restore.go:978]: Will most of these case be catched by `!common.IsContextCanceledError(err)` ?

Comment (D3Hunter) [inline br/pkg/lightning/restore/restore.go:978]: this case should move before `case !common.IsContextCanceledError(err):`, else it's never run

File: br/pkg/lightning/restore/restore.go [hunk 3]
```diff
@@ -975,6 +975,11 @@ func (rc *Controller) saveStatusCheckpoint(ctx context.Context, tableName string
...
+		// recoverable error, should not be recorded in checkpoint
+		// which will prevent lightning from automatically recovering
+		rc.errorSummaries.record(tableName, err, statusIfSucceed)
```
Comment (D3Hunter) [inline br/pkg/lightning/restore/restore.go:981]: should we record this error? since it's recoverable

errorSummaries is emitted after `controller.Run`.

**Review by D3Hunter — APPROVED** [review]

**Review by gozssky — APPROVED** [review]

**Review by D3Hunter — COMMENTED** [review]
lgtm

**Pull request merged**